libraries/provider_systemd_unit.rb

```diff
@@ -14,36 +14,35 @@ def whyrun_supported?
       provides "systemd_#{unit_type}".to_sym
     end
 
-    action :create do
-      unit_path = Systemd::Helpers.unit_path(new_resource)
-
-      directory Systemd::Helpers.drop_in_root(new_resource) do
-        only_if { new_resource.drop_in }
-      end
-
-      execute 'reload-sytemd' do
-        command 'systemctl daemon-reload'
-        action :nothing
-        subscribes :run, "file[#{unit_path}]", :immediately
-      end
-
-      file unit_path do
-        content Systemd::Helpers.ini_config(new_resource.to_hash)
-        action :create
+    %i( create delete ).each do |a|
+      action a do
+        r = new_resource
+        unit_path = Systemd::Helpers.unit_path(r)
+
+        directory Systemd::Helpers.drop_in_root(r) do
+          only_if { r.drop_in }
+          not_if { r.action == :delete }
+        end
+
+        execute "#{r.name}.#{r.unit_type}-systemd-reload" do
+          command 'systemctl daemon-reload'
+          action :nothing
+          subscribes :run, "file[#{unit_path}]", :immediately
+        end
+
+        file unit_path do
+          content Systemd::Helpers.ini_config(r.to_hash)
+          action r.action
+        end
       end
     end
 
-    action :delete do
-      unit_path = Systemd::Helpers.unit_path(new_resource)
-
-      execute 'reload-sytemd' do
-        command 'systemctl daemon-reload'
-        action :nothing
-        subscribes :run, "file[#{unit_path}]", :immediately
-      end
-
-      file unit_path do
-        action :delete
+    %i( enable disable start stop ).each do |a|
+      action a do
+        r = new_resource
+        execute "systemctl-#{a}-#{r.name}.#{r.unit_type}" do
+          command "systemctl #{a} #{r.name}.#{r.unit_type}"
```

Comment:
yeah, i wasn't quite sure about that... systemd seems to handle the idempotence itself, so if the service is already in the specified state, executing the command is a no-op on the system.

```bash
[root@default-centos-71 ~]# systemctl status rsyslog.service
rsyslog.service - System Logging Service
   Loaded: loaded (/usr/lib/systemd/system/rsyslog.service; enabled)
   Active: active (running) since Fri 2015-07-10 15:15:03 UTC; 1min 44s ago
 Main PID: 543 (rsyslogd)
   CGroup: /system.slice/rsyslog.service
           └─543 /usr/sbin/rsyslogd -n

Jul 10 15:15:03 localhost.localdomain systemd[1]: Starting System Logging Service...
Jul 10 15:15:03 localhost.localdomain systemd[1]: Started System Logging Service.
[root@default-centos-71 ~]# systemctl stop rsyslog.service
[root@default-centos-71 ~]# systemctl disable rsyslog.service
rm '/etc/systemd/system/multi-user.target.wants/rsyslog.service'
[root@default-centos-71 ~]# echo $?
0
[root@default-centos-71 ~]# systemctl disable rsyslog.service
[root@default-centos-71 ~]# echo $?
0
[root@default-centos-71 ~]# systemctl stop rsyslog.service
[root@default-centos-71 ~]# echo $?
0
[root@default-centos-71 ~]# systemctl status rsyslog.service
rsyslog.service - System Logging Service
   Loaded: loaded (/usr/lib/systemd/system/rsyslog.service; disabled)
   Active: inactive (dead)

Jul 10 15:15:03 localhost.localdomain systemd[1]: Starting System Logging Service...
Jul 10 15:15:03 localhost.localdomain systemd[1]: Started System Logging Service.
Jul 10 15:16:57 default-centos-71 systemd[1]: Stopping System Logging Service...
Jul 10 15:16:57 default-centos-71 systemd[1]: Stopped System Logging Service.
Jul 10 15:17:26 default-centos-71 systemd[1]: Stopped System Logging Service.
[root@default-centos-71 ~]# systemctl enable rsyslog.service
ln -s '/usr/lib/systemd/system/rsyslog.service' '/etc/systemd/system/multi-user.target.wants/rsyslog.service'
[root@default-centos-71 ~]# echo $?
0
[root@default-centos-71 ~]# systemctl enable rsyslog.service
[root@default-centos-71 ~]# echo $?
0
[root@default-centos-71 ~]# systemctl start rsyslog.service
[root@default-centos-71 ~]# echo $?
0
[root@default-centos-71 ~]# systemctl start rsyslog.service
[root@default-centos-71 ~]# echo $?
0
[root@default-centos-71 ~]# systemctl status rsyslog.service
rsyslog.service - System Logging Service
   Loaded: loaded (/usr/lib/systemd/system/rsyslog.service; enabled)
   Active: active (running) since Fri 2015-07-10 15:17:48 UTC; 18s ago
 Main PID: 12403 (rsyslogd)
   CGroup: /system.slice/rsyslog.service
           └─12403 /usr/sbin/rsyslogd -n

Jul 10 15:17:48 default-centos-71 systemd[1]: Starting System Logging Service...
Jul 10 15:17:48 default-centos-71 systemd[1]: Started System Logging Service.
Jul 10 15:18:02 default-centos-71 systemd[1]: Started System Logging Service.
```

since we're not controlling updated_by_last_action, i wasn't sure what the effect here would be on the resource (thinking about it now in the light of day, i'm guessing it would always consider itself updated?), and knowing it was safe to repeat the command, went with the ~~laziest~~ simplest implementation to start.

i'll see about switching to mixlib shellout and see if we can make the resource triggering more conditional on the state so the resource updated_by_last_action? can be trusted.